src/shell/shell.cpp

```diff
@@ -178,141 +178,283 @@ DOS_Shell::DOS_Shell()
           call(false)
 {}
 
-// TODO: this function should be refactored to make to it easier to understand.
-// It's currently riddled with pointer and array adjustments each loop plus
-// branches and sub-loops.
-Bitu DOS_Shell::GetRedirection(char *s, char **ifn, char **ofn, bool *append)
+void DOS_Shell::GetRedirection(char *line,
+                               std::string &in_file,
+                               std::string &out_file,
+                               std::string &pipe_file,
+                               bool *append)
```

Comment:
No C-strings, references instead of double-pointers, and self-documenting names w/ directions (in/out). 
This is a huge improvement.